src/smt/theory_str_noodler/theory_str_noodler.cpp

```diff
@@ -1960,6 +1975,136 @@ namespace smt::noodler {
+
+        // get the var for the argument
+        expr_ref var_for_s(m);
+        if (tranforming_from) {
+            // for from_code and from_int, the argument has integer type, we create a new var for it
+            var_for_s = mk_int_var_fresh(name_of_type + "_argument");
+            // bit of a hack, we abuse var_name to save the predicate to which int var is mapped, because
+            // using equation here did not work (lia solver ignored it)
+            var_name.insert({util::get_variable_basic_term(var_for_s), expr_ref(s, m)});
+        } else {
+            // for to_code and to_int, the argument has string type, we have to find the variable for it
+            if (m_util_s.str.is_string(s)) {
+                // it seems that Z3 rewriter handles the case where we tranform from string literal, so this should be unreachable
+                UNREACHABLE();
+                return;
+            } else if (util::is_str_variable(s, m_util_s)) {
+                // we are transforming directly from variable
+                var_for_s = s;
+            } else if(this->predicate_replace.contains(s)) {
+                // argument is some function that already has a replacing variable
+                var_for_s = this->predicate_replace[s];
+            } else {
+                // argument does not have a replacing variable (probably concatenation)
+                // we need to create one
+                var_for_s = mk_str_var_fresh(name_of_type + "_argument");
+                add_axiom({mk_literal(m.mk_eq(s, var_for_s))});
+                this->predicate_replace.insert(s, var_for_s);
+            }
+            // we need exact solution for the argument, so that we can compute
+            // the arithmetic formula for the result in final_check_eh
+            len_vars.insert(var_for_s);
+        }
+
+        // create var for the result
+        expr_ref var_for_e(m);
+        if (tranforming_from) {
+            var_for_e = mk_str_var_fresh(name_of_type + "_result");
+            add_axiom({mk_literal(m.mk_eq(var_for_e, e))});
+            this->predicate_replace.insert(e, var_for_e);
+            len_vars.insert(var_for_e); // dunno if this is needed
+        } else {
+            var_for_e = mk_int_var_fresh(name_of_type + "_result");
+            // bit of a hack, we abuse var_name to save the predicate to which int var is mapped, because
+            // using equation here did not work (lia solver ignored it)
+            var_name.insert({util::get_variable_basic_term(var_for_e), expr_ref(e, m)});
+        }
+
+        // The range of from_* functions is bounded, we have to bound it also for the decision procedure
+
+        if (type == TransformationType::FROM_CODE) {
+            // the result of str.from_code can only be either a char representing the code value, or empty string (if argument is out of range of any code value)
+            app *sigma_eps = m_util_s.re.mk_union(
```

Comment:
You `app_ref` instead? Doesn't it cause segfault? Maybe not as it is used inside `mk_literal`.

Reply:
I did not know `app_ref` exists. Or did you mean `expr_ref`? I have no idea why it does not cause segfault, but I can change it, if you think it is safer.

Reply:
If you tried it and it works, it is fine.